### PR TITLE
Cleanup pipelines code and tests

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: samples/System.IO.Pipelines.Samples/CompressionSample.cs [hunk 1]
```diff
@@ -14,38 +14,35 @@ public class CompressionSample : ISample
     {
         public Task Run()
         {
-            using (var bufferPool = new MemoryPool())
-            {
-                var filePath = Path.GetFullPath("Program.cs");
+            var filePath = Path.GetFullPath("Program.cs");
```
Comment (ahsonkhan) [inline samples/System.IO.Pipelines.Samples/CompressionSample.cs]: Use of var here and elsewhere

File: src/System.Buffers.Primitives/System/Buffers/IOutput.cs [hunk 2]
```diff
@@ -3,12 +3,26 @@
 
 namespace System.Buffers
 {
+    /// <summary>
+    /// Represents a <see cref="Byte"/> sink
+    /// </summary>
     public interface IOutput
     {
+        /// <summary>
+        /// Notifies <see cref="IOutput"/> that <see cref="bytes"/> amount of data was written to <see cref="Span{Byte}"/>/<see cref="Memory{Byte}"/>
```
Comment (ahsonkhan) [inline src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: paramref instead of cref for "bytes"

Here and elsewhere.
I believe the pattern is to use paramref for parameter names, and cref for types, methods, and properties.

File: src/System.Buffers.Primitives/System/Buffers/IOutput.cs [hunk 3]
```diff
@@ -3,12 +3,26 @@
...
+        /// </summary>
         void Advance(int bytes);
-        /// <summary>If minimumLength is equal to zero currently available memory would be returned</summary>
+
+        /// <summary>
+        /// Requests the <see cref="Memory{Byte}"/> of at least <see cref="minimumLength"/> in size.
+        /// If <see cref="minimumLength"/> is equal to <code>0</code> currently available memory would get returned.
```
Comment (ahsonkhan) [inline src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: nit: add comma, `is equal to <code>0</code>, currently available memory would get returned.`

File: src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs [hunk 4]
```diff
@@ -351,7 +351,7 @@ private enum BufferStyle
                         // before flushing (entirely arbitrarily) - might want to make this configurable later
 
                         haveWriteBuffer = true;
-                        int FlushInputEveryBytes = 4 * Pool.MaxBufferSize;
+                        int FlushInputEveryBytes = 24;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs]: Can you please explain why this change was needed?

File: src/System.IO.Pipelines/System/IO/Pipelines/FlushResult.cs [hunk 5]
```diff
@@ -23,12 +30,12 @@ public FlushResult(bool isCancelled, bool isCompleted)
         }
 
         /// <summary>
-        /// True if the currrent flush was cancelled
+        /// True if the current <see cref="PipeWriter.FlushAsync"/> operation was canceled otherwise false
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/System/IO/Pipelines/FlushResult.cs]: nit: Change to one of (here and elsewhere):
`True if the current <see cref="PipeWriter.FlushAsync"/> operation was canceled, otherwise false.`
`True if the current <see cref="PipeWriter.FlushAsync"/> operation was canceled. Otherwise false.`

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs [hunk 6]
```diff
@@ -46,18 +47,18 @@ public abstract class PipeReader
         public abstract void AdvanceTo(SequencePosition consumed, SequencePosition examined);
 
         /// <summary>
-        /// Cancel to currently pending or next call to <see cref="ReadAsync"/> if none is pending, without completing the <see cref="PipeReader"/>.
+        /// Cancel to currently pending or if none is pending next call to <see cref="ReadAsync"/>, without completing the <see cref="PipeReader"/>.
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs]: Re-write comment.

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs [hunk 7]
```diff
@@ -46,18 +47,18 @@ public abstract class PipeReader
...
         /// <summary>
         /// Signal to the producer that the consumer is done reading.
         /// </summary>
-        /// <param name="exception">Optional Exception indicating a failure that's causing the pipeline to complete.</param>
+        /// <param name="exception">Optional <see cref="Exception"/> indicating a failure that's causing the pipeline to complete.</param>
         public abstract void Complete(Exception exception = null);
 
         /// <summary>
-        /// Registers callback that gets executed when writer side of pipe completes.
+        /// Registers callback that gets executed when <see cref="PipeWriter"/> side of pipe complete.
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs]: `Registers a callback that gets executed when the <see cref="PipeWriter"/> side of the pipe finishes.`

File: src/System.IO.Pipelines/System/IO/Pipelines/ReadResult.cs [hunk 8]
```diff
@@ -22,21 +25,21 @@ public ReadResult(ReadOnlyBuffer<byte> buffer, bool isCancelled, bool isComplete
...
+        /// The <see cref="ReadOnlyBuffer{Byte}"/> that was read
         /// </summary>
         public ReadOnlyBuffer<byte> Buffer => ResultBuffer;
 
         /// <summary>
-        /// True if the currrent read was cancelled
+        /// True if the current <see cref="PipeReader.ReadAsync"/> operation was canceled
         /// </summary>
-        public bool IsCancelled => (ResultFlags & ResultFlags.Cancelled) != 0;
+        public bool IsCanceled => (ResultFlags & ResultFlags.Cancelled) != 0;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/System/IO/Pipelines/ReadResult.cs]: Inconsistency - we have Cancelled in some places and Canceled in other places.

File: src/System.IO.Pipelines/System/Threading/PipeScheduler.cs [hunk 9]
```diff
@@ -1,17 +1,34 @@
...
+        /// <summary>
+        /// The <see cref="PipeScheduler"/> implementation that runs callbacks inline
+        /// </summary>
         public static PipeScheduler Inline => _inlineScheduler;
 
+        /// <summary>
+        /// Requests <see cref="action"/> to be run on scheduler
+        /// </summary>
         public abstract void Schedule(Action action);
+        /// <summary>
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/System/Threading/PipeScheduler.cs]: nit: add a new line

File: tests/Benchmarks/Helpers/Server.cs [hunk 10]
```diff
@@ -67,7 +66,6 @@ public static void Run(int numberOfRequests, int concurrentConnections, byte[] r
             Task.WaitAll(tasks);
 
             listener.Dispose();
-            memoryPool.Dispose();
```
Comment (ahsonkhan) [inline tests/Benchmarks/Helpers/Server.cs]: We no longer have to call dispose on memorypool? Is it because of the change to use the Shared MemoryPool rather than instantiating a new one? cc @AtsushiKan

File: tests/System.IO.Pipelines.Tests/PipeTest.cs [hunk 11]
```diff
@@ -9,16 +10,19 @@ public abstract class PipeTest : IDisposable
     {
         protected const int MaximumSizeHigh = 65;
 
+        private readonly TestMemoryPool _pool;
+
         protected Pipe Pipe;
-        private readonly MemoryPool _pool;
 
-        protected PipeTest()
+        protected PipeTest(int pauseWriterThreshold = 65, int resumeWriterThreshold = 6)
```
Comment (ahsonkhan) [inline tests/System.IO.Pipelines.Tests/PipeTest.cs]: nit: Use MaximumSizeHigh?

File: src/System.IO.Pipelines/System/Buffers/MemoryPoolBlock.cs [hunk 12]
```diff
@@ -1,147 +0,0 @@
...
-using System.Diagnostics;
-using System.IO.Pipelines;
-
-namespace System.Buffers
-{
-    /// <summary>
-    /// Block tracking object used by the byte buffer memory pool. A slab is a large allocation which is divided into smaller blocks. The
-    /// individual blocks are then treated as independent array segments.
-    /// </summary>
-    internal class MemoryPoolBlock : OwnedMemory<byte>
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/System/Buffers/MemoryPoolBlock.cs]: Why don't we need this and MemoryPoolSlab anymore?

**Comment by ahsonkhan** [comment]
cc @KrzysztofCwalina

File: samples/System.IO.Pipelines.Samples/CompressionSample.cs (hunk 1 above)
Comment (pakrym) [reply samples/System.IO.Pipelines.Samples/CompressionSample.cs]: ?

File: src/System.IO.Pipelines/System/Buffers/MemoryPoolBlock.cs (hunk 12 above)
Comment (pakrym) [reply src/System.IO.Pipelines/System/Buffers/MemoryPoolBlock.cs]: We are not shipping this implementation in corefx, only in kestrel

File: tests/System.IO.Pipelines.Tests/PipeTest.cs (hunk 11 above)
Comment (pakrym) [reply tests/System.IO.Pipelines.Tests/PipeTest.cs]: why?

File: samples/System.IO.Pipelines.Samples/CompressionSample.cs (hunk 1 above)
Comment (ahsonkhan) [reply samples/System.IO.Pipelines.Samples/CompressionSample.cs]: State type explicitly where it isn't statically clear from the right hand side of the expression, what the type is. Only use vars in those scenarios.

File: tests/System.IO.Pipelines.Tests/PipeTest.cs (hunk 11 above)
Comment (ahsonkhan) [reply tests/System.IO.Pipelines.Tests/PipeTest.cs]: We have a constant already. I assume this 65 has the same semantic meaning. If not, can you add a comment why these constant values are special?

Comment (pakrym) [reply tests/System.IO.Pipelines.Tests/PipeTest.cs]: Ah, my bad, you are completely right.

File: samples/System.IO.Pipelines.Samples/CompressionSample.cs (hunk 1 above)
Comment (pakrym) [reply samples/System.IO.Pipelines.Samples/CompressionSample.cs]: Use of var, I read "Use var". Will do.

**Comment by pakrym** [comment]
@ahsonkhan why aren't CI build running?

**Comment by ahsonkhan** [comment]
> why aren't CI build running?

Not sure. Let me try re-starting it.

@dotnet-bot test this please

@dotnet/dnceng might know.

**Comment by ahsonkhan** [comment]
OK. It is working now.

File: src/System.Buffers.Primitives/System/Buffers/IOutput.cs [hunk 13]
```diff
@@ -3,12 +3,26 @@
 
 namespace System.Buffers
 {
+    /// <summary>
+    /// Represents a <see cref="Byte"/> sink
+    /// </summary>
     public interface IOutput
```
Comment (davidfowl) [inline src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: Rename to IBufferWriter

File: src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs (hunk 4 above)
Comment (pakrym) [reply src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs]: Yeah, it's a random flush size computation in the sample and it overflowed when `Pool.MaxBufferSize` is maxint

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs (hunk 7 above)
Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs]: completes is here because we have Complete method and want to show connection

File: tests/Benchmarks/Helpers/Server.cs (hunk 10 above)
Comment (pakrym) [reply tests/Benchmarks/Helpers/Server.cs]: Yes, it's because we don't create it anymore.

**Comment by pakrym** [comment]
> because of the indentation updates.

@ahsonkhan you can tell github to ignore whitespace (https://github.com/dotnet/corefxlab/pull/2103/files?w=1)

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs (hunk 7 above)
Comment (ahsonkhan) [reply src/System.IO.Pipelines/System/IO/Pipelines/PipeReader.cs]: Sounds good. Change to `completes` (plural):
`Registers a callback that gets executed when the <see cref="PipeWriter"/> side of the pipe completes.`

**Comment by ahsonkhan** [comment]
> @ahsonkhan you can tell github to ignore whitespace (https://github.com/dotnet/corefxlab/pull/2103/files?w=1)

Awesome. Thanks for reminding me. I had forgotten (https://github.com/dotnet/corefxlab/pull/1636#issuecomment-311464942).
